source/common/http/conn_manager_impl.cc

```diff
@@ -470,6 +488,62 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
       connection_manager_.config_, *snapped_route_config_, connection_manager_.random_generator_,
       connection_manager_.runtime_, connection_manager_.local_info_);
 
+  if (!cached_route_.valid()) {
+    cached_route_.value(snapped_route_config_->route(*request_headers_, stream_id_));
+  }
+
+  // Check for WebSocket upgrade request if the route is valid, and supports WebSockets.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
+    const Router::RouteEntry* route_entry = cached_route_.value()->routeEntry();
+    // Websocket route entries cannot be redirects.
+    bool websocket_upgrade_allowed =
+        (route_entry != nullptr) && route_entry->allowWebSocketUpgrade();
+    bool websocket_upgrade_requested = Utility::isWebSocketUpgradeRequest(*request_headers_);
+
+    if (websocket_upgrade_requested && websocket_upgrade_allowed) {
+      ENVOY_STREAM_LOG(debug, "found websocket connection. (end_stream={}):", *this, end_stream);
+
+      // A very kludgey way to get a handle over the stream
+      // decoder! WsHandlerImpl needs the stream decoder in order to
+      // properly terminate connections with HTTP error codes when
+      // there is an error establishing connection to upstream.
+      // TODO (rshriram) - can decoder_filters_ be empty? Is there a better way to get
+      // a handle over the stream decoder filter callback instead of ** ?
+      std::list<ActiveStreamDecoderFilterPtr>::iterator entry = decoder_filters_.begin();
+      if (entry != decoder_filters_.end()) {
+        connection_manager_.ws_connection_ =
+            std::unique_ptr<WebSocket::WsHandlerImpl>(new WebSocket::WsHandlerImpl(
```

Comment:
Can you avoid some of this kludgeyness and not use an existing filter to do encodeHeaders() by doing something like this:

1) Make a new interface, something like WebsocketHandlerCallbacks, with a method like sendHeadersOnlyResponse() - this is nice because you can mock it for tests.

2) Have your implementation of WebsocketHandlerCallbacks hold the WsHandlerImpl.

3) Connection Manager can instantiate WebsocketHandlerCallbacks impl, get headers with the sendHeadersOnlyResponse() method, call encodeHeaders, and then forward to the parent active stream.